packages/build/src/plugins_core/deploy_config/index.ts

```diff
@@ -0,0 +1,87 @@
+import { promises as fs } from 'fs'
+import { resolve } from 'path'
+
+import merge from 'deepmerge'
+
+import type { NetlifyConfig } from '../../index.js'
+import { CoreStep, CoreStepFunction } from '../types.js'
+
+import { filterConfig } from './util.js'
+
+const ALLOWED_PROPERTIES = [
+  ['build', 'edge_functions'],
+  ['build', 'functions'],
+  ['build', 'publish'],
+  ['functions', '*'],
+  ['functions', '*', '*'],
+  ['headers'],
+  ['images', 'remote_images'],
+  ['redirects'],
+]
+
+const OVERRIDE_PROPERTIES = new Set(['headers!', 'redirects!'])
+
+const coreStep: CoreStepFunction = async function ({
+  buildDir,
+  netlifyConfig,
+  systemLog = () => {
+    // no-op
+  },
+}) {
+  const configPath = resolve(buildDir, '.netlify', 'deploy', 'v1', 'config.json')
+
+  let config: Partial<NetlifyConfig> = {}
+
+  try {
+    const data = await fs.readFile(configPath, 'utf8')
+
+    config = JSON.parse(data) as Partial<NetlifyConfig>
+  } catch (err) {
+    // If the file doesn't exist, this is a non-error.
+    if (err.code === 'ENOENT') {
+      return {}
+    }
+
+    systemLog(`Failed to read Deploy Configuration API: ${err.message}`)
+
+    throw new Error('An error occured while processing the platform configurarion defined by your framework.')
+  }
+
+  const configOverrides: Partial<NetlifyConfig> = {}
+
+  for (const key in config) {
+    // If the key uses the special notation for defining mutations that should
+    // take precedence over user-defined properties, extract the canonical
+    // property, set it on a different object, and delete it from the main one.
+    if (OVERRIDE_PROPERTIES.has(key)) {
+      const canonicalKey = key.slice(0, -1)
+
+      configOverrides[canonicalKey] = config[key]
+
+      delete config[key]
+    }
+  }
+
+  // Filtering out any properties that can't be mutated using this API.
+  config = filterConfig(config, [], ALLOWED_PROPERTIES, systemLog)
```

Comment:
Is the reason for us to do this here too just fail fast vs letting the plugin run and ntl/config trying to run the mutations?

Reply:
Also it would be useful to maybe add some system logging or span event saying we've skipped X properties because we're filtering or something like that.

Reply:
> Is the reason for us to do this here too just fail fast vs letting the plugin run and ntl/config trying to run the mutations?

I think the Deploy Configuration API should have its own set of properties that can be mutated — for example, `build.command` can be mutated by a plugin, but it doesn't make sense for it to be mutated using the Deploy Configuration API since that is expected to be generated during the build command.

Reply:
> Also it would be useful to maybe add some system logging or span event saying we've skipped X properties because we're filtering or something like that.

I've added some system logging for every property we use and discard.